consensus/istanbul/backend/backend.go

```diff
@@ -939,3 +957,24 @@ func (sb *Backend) VerifyValidatorConnectionSetSignature(data []byte, sig []byte
 		return istanbul.CheckValidatorSignature(validator.NewSet(validators), data, sig)
 	}
 }
+
+func (sb *Backend) UpdateReplicaState() bool {
```

Comment:
It seems to me that there's this "state machine" where we have several inputs:
```
-> newBlockHeader
-> makePrimary/Replica
-> start/stopXXX
-> configureStop/StartBlock
```

and those action might generate state transtitions, which trigger start or stop of subservices.

Now, today, that logic is dispersed in a few methods: `UpdateReplicatState` `Start/StopValidating` `MakeReplica/Primary`. I wonder if it would be better to have the control logic isolated on its own entity, and assign some trigger function of "start/stopXXX" when you enter/exit states.

That way, control logic would be clearer.

Reply:
There definitely is. I'd put the inputs as follows:
```
-> block number (from bc.CurrentBlock() or from roundstate.Seq)
-> start validating (external RPC, also clears start/stop blocks)
-> stop validating (external RPC, also clears start/stop blocks)
-> set start validating block (external RPC)
-> set stop validating block (external RPC)
```
One thing I was grappling with is the outputs
```
-> Start core validating
-> Stop core from validating
```
Both of those are function calls, so it's figuring out where this can live and have access to those functions in a clean way.